v2/yarpcthrift/outbound.go

```diff
@@ -0,0 +1,221 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package yarpcthrift
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+
+	"go.uber.org/thriftrw/envelope"
+	"go.uber.org/thriftrw/protocol"
+	"go.uber.org/thriftrw/wire"
+	"go.uber.org/yarpc/v2"
+	"go.uber.org/yarpc/v2/yarpcencoding"
+	"go.uber.org/yarpc/v2/yarpcprocedure"
+	"go.uber.org/yarpc/v2/yarpcthrift/internal"
+)
+
+// Client is a generic Thrift client. It speaks in raw Thrift payloads.
+//
+// Users should use the client generated by the code generator rather than
+// using this directly.
+type Client interface {
+	// Call the given Thrift method.
+	Call(ctx context.Context, reqBody envelope.Enveloper, opts ...yarpc.CallOption) (wire.Value, error)
+}
+
+// Config contains the configuration for the Client.
+type Config struct {
+	// Name of the Thrift service. This is the name used in the Thrift file
+	// with the 'service' keyword.
+	Service string
+
+	// ClientConfig through which requests will be sent. Required.
+	Client *yarpc.Client
+}
+
+// New creates a new Thrift client.
+func New(c Config, opts ...ClientOption) Client {
+	// Code generated for Thrift client instantiation will probably be something
+	// like this:
+	//
+	// 	func New(tc thrift.Client, opts ...ClientOption) *MyServiceClient {
+	// 		c := thrift.New(thrift.Config{
+	// 			Service: "MyService",
+	// 			Client: tc,
+	// 		}, opts...)
+	// 		return &MyServiceClient{client: c}
+	// 	}
+	//
+	// So Config is really the internal config as far as consumers of the
+	// generated client are concerned.
+
+	var cc clientConfig
+	for _, opt := range opts {
+		opt.applyClientOption(&cc)
+	}
+
+	p := protocol.Binary
+	if cc.Protocol != nil {
+		p = cc.Protocol
+	}
+
+	if cc.Multiplexed {
+		p = multiplexedOutboundProtocol{
+			Protocol: p,
+			Service:  c.Service,
+		}
+	}
+
+	return thriftClient{
+		p:             p,
+		c:             c.Client,
+		thriftService: c.Service,
+		Enveloping:    cc.Enveloping,
+	}
+}
+
+type thriftClient struct {
+	c *yarpc.Client
+	p protocol.Protocol
+
+	// name of the Thrift service
+	thriftService string
+	Enveloping    bool
+}
+
+func (c thriftClient) Call(ctx context.Context, reqBody envelope.Enveloper, opts ...yarpc.CallOption) (wire.Value, error) {
+	// Code generated for Thrift client calls will probably be something like
+	// this:
```

Comment:
We should make a note to revise or erase this comment before finalizing v2, in CHANGES. This comment speculates about the final implementation. I agree there are some pending choices!

Reply:
looks like you missed this one @mh-park, removing it, as Kris suggests is fine.